File: src/accessiweather/display/presentation/alerts.py
"""Builders for weather alert presentations."""

from __future__ import annotations

from typing import TYPE_CHECKING

from ...models import AppSettings, Location, WeatherAlert, WeatherAlerts
from ..weather_presenter import AlertPresentation, AlertsPresentation
from .formatters import format_display_datetime, truncate, wrap_text

if TYPE_CHECKING:
    from accessiweather.alert_lifecycle import AlertLifecycleDiff


def _is_pirate_weather_alert(alert: WeatherAlert) -> bool:
    """Return True when the alert originated from Pirate Weather / WMO."""
    return "pirateweather" in (alert.source or "").strip().lower()


def build_alerts(
    alerts: WeatherAlerts,
    location: Location,
    settings: AppSettings | None = None,
    *,
    lifecycle_diff: AlertLifecycleDiff | None = None,
    lifecycle_states: dict[str, str] | None = None,
) -> AlertsPresentation:
    """
    Create an alerts presentation for a given location.

    Args:
        alerts: Current weather alerts snapshot.
        location: Location being presented.
        settings: Optional app settings for time formatting.
        lifecycle_diff: Optional diff from the previous alert snapshot.
            When provided and has changes, a concise change summary is
            prepended to the fallback text and stored on the presentation.
        lifecycle_states: Optional mapping of alert_id to label (e.g.
            {"alert-1": "New", "alert-2": "Extended"}).  When provided,
            labels are appended to each alert's header line in the
            presentation (e.g. "Alert 1: Dense Fog Advisory (Extended)").

    """
    title = f"Weather alerts for {location.name}"
    if not alerts.has_alerts():
        return AlertsPresentation(title=title, fallback_text=f"{title}:\nNo active weather alerts.")

    active = alerts.get_active_alerts()
    if not active:
        return AlertsPresentation(title=title, fallback_text=f"{title}:\nNo active weather alerts.")

    presentations: list[AlertPresentation] = []
    fallback_lines = [title + ":"]

    for idx, alert in enumerate(active, start=1):
        lifecycle_label: str | None = None
        if lifecycle_states:
            alert_id = alert.get_unique_id()
            lifecycle_label = lifecycle_states.get(alert_id)
        presentation = build_single_alert(alert, idx, settings, lifecycle_label=lifecycle_label)
        presentations.append(presentation)
        fallback_lines.append(presentation.fallback_text)

    fallback_text = "\n\n".join(fallback_lines)

    change_summary: str | None = None
    if lifecycle_diff is not None and lifecycle_diff.has_changes:
        change_summary = lifecycle_diff.summary
        fallback_text = f"Alert changes: {lifecycle_diff.summary}\n{fallback_text}"

    return AlertsPresentation(
        title=title,
        alerts=presentations,
        fallback_text=fallback_text,
        change_summary=change_summary,
    )


def build_single_alert(
    alert: WeatherAlert,
    index: int,
    settings: AppSettings | None = None,
    *,
    lifecycle_label: str | None = None,
) -> AlertPresentation:
    """
    Create a single alert presentation with truncated text for readability.

    Args:
        alert: The weather alert to present.
        index: 1-based display index.
        settings: Optional app settings for time formatting.
        lifecycle_label: Optional lifecycle status label (e.g. "New", "Updated",
            "Escalated", "Extended").  When provided it is appended to the header
            line: "Alert 1: Dense Fog Advisory (Extended)".

    """
    severity = alert.severity if alert.severity != "Unknown" else None
    urgency = alert.urgency if alert.urgency != "Unknown" else None
    areas = alert.areas[:3] if alert.areas else []
    is_regional_pw_alert = _is_pirate_weather_alert(alert)

    # Extract time preferences
    if settings:
        time_display_mode = getattr(settings, "time_display_mode", "local")
        time_format_12hour = getattr(settings, "time_format_12hour", True)
        show_timezone_suffix = getattr(settings, "show_timezone_suffix", False)
    else:
        time_display_mode = "local"
        time_format_12hour = True
        show_timezone_suffix = False

    expires = None
    if alert.expires:
        expires = format_display_datetime(
            alert.expires,
            time_display_mode=time_display_mode,
            use_12hour=time_format_12hour,
            show_timezone=show_timezone_suffix,
            date_format="%m/%d",
        )

    description = truncate(alert.description, 200) if alert.description else None
    instruction = truncate(alert.instruction, 150) if alert.instruction else None

    alert_header = f"Alert {index}: {alert.title}" if alert.title else f"Alert {index}"
    if lifecycle_label:
        alert_header = f"{alert_header} ({lifecycle_label})"
    parts: list[str] = [alert_header]
    if severity or urgency:
        sev_bits = []
        if severity:
            sev_bits.append(f"Severity: {severity}")
        if urgency:
            sev_bits.append(f"Urgency: {urgency}")
        parts.append("  " + ", ".join(sev_bits))
    if alert.event:
        parts.append(f"  Event: {alert.event}")
    if areas:
        remaining = len(alert.areas) - len(areas)
        area_text = ", ".join(areas)
        if remaining > 0:
            area_text += f" and {remaining} more"
        area_label = "Regions" if is_regional_pw_alert else "Areas"
        parts.append(f"  {area_label}: {area_text}")
    if is_regional_pw_alert:
        parts.append(
            "  Coverage: Regional alert from Pirate Weather/WMO; may not match your exact county or zone."
        )
    if expires:
        parts.append(f"  Expires: {expires}")
    if description:
        parts.append(f"  Description: {wrap_text(description, 80)}")
    if instruction:
        parts.append(f"  Instructions: {wrap_text(instruction, 80)}")

    fallback_text = "\n".join(parts)
    return AlertPresentation(
        title=alert.title or alert.event or f"Alert {index}",
        severity=severity,
        urgency=urgency,
        event=alert.event,
        areas=alert.areas or [],
        expires=expires,
        description=description,
        instructions=instruction,
        fallback_text=fallback_text,
    )

File: rust/crates/aw-core/src/display/alerts.rs
//! Weather alert presentation.
//!
//! Port of `display/presentation/alerts.py`.

use std::collections::HashMap;

use chrono::{DateTime, Utc};
use chrono_tz::Tz;

use crate::display::models::{AlertPresentation, AlertsPresentation};
use crate::display::pyfmt::{truncate, wrap_text};
use crate::display::time::{format_display_datetime, PyDateTime};
use crate::model::{AlertLifecycleDiff, WeatherAlert, WeatherAlerts};
use crate::settings::AppSettings;

/// `AlertLifecycleDiff.has_changes`.
pub fn lifecycle_has_changes(diff: &AlertLifecycleDiff) -> bool {
    !(diff.new_alerts.is_empty()
        && diff.updated_alerts.is_empty()
        && diff.escalated_alerts.is_empty()
        && diff.extended_alerts.is_empty()
        && diff.cancelled_alerts.is_empty())
}

/// `WeatherAlerts.get_active_alerts`: no expiry, or expiring after `now`.
pub fn active_alerts(alerts: &WeatherAlerts, now: DateTime<Utc>) -> Vec<&WeatherAlert> {
    alerts
        .alerts
        .iter()
        .filter(|a| a.expires.is_none_or(|e| e > now))
        .collect()
}

fn is_pirate_weather_alert(alert: &WeatherAlert) -> bool {
    alert
        .source
        .as_deref()
        .unwrap_or("")
        .trim()
        .to_lowercase()
        .contains("pirateweather")
}

/// `build_alerts`. `lifecycle_states` maps `unique_id()` to a label such as
/// "Extended" that is appended to the alert's header line.
pub fn build_alerts(
    alerts: &WeatherAlerts,
    location_name: &str,
    settings: &AppSettings,
    location_zone: Option<Tz>,
    lifecycle_diff: Option<&AlertLifecycleDiff>,
    lifecycle_states: Option<&HashMap<String, String>>,
    now: DateTime<Utc>,
) -> AlertsPresentation {
    let title = format!("Weather alerts for {location_name}");
    let active = active_alerts(alerts, now);
    if active.is_empty() {
        return AlertsPresentation {
            fallback_text: format!("{title}:\nNo active weather alerts."),
            title,
            ..Default::default()
        };
    }

    let mut presentations = Vec::new();
    let mut fallback_lines = vec![format!("{title}:")];
    for (idx, alert) in active.into_iter().enumerate() {
        let label = lifecycle_states
            .filter(|m| !m.is_empty())
            .and_then(|m| m.get(&alert.unique_id()))
            .map(String::as_str);
        let p = build_single_alert(alert, idx + 1, settings, location_zone, label);
        fallback_lines.push(p.fallback_text.clone());
        presentations.push(p);
    }
    let mut fallback_text = fallback_lines.join("\n\n");

    let mut change_summary = None;
    if let Some(diff) = lifecycle_diff.filter(|d| lifecycle_has_changes(d)) {
        change_summary = Some(diff.summary.clone());
        fallback_text = format!("Alert changes: {}\n{fallback_text}", diff.summary);
    }

    AlertsPresentation {
        title,
        alerts: presentations,
        fallback_text,
        change_summary,
    }
}

/// `build_single_alert`.
pub fn build_single_alert(
    alert: &WeatherAlert,
    index: usize,
    settings: &AppSettings,
    location_zone: Option<Tz>,
    lifecycle_label: Option<&str>,
) -> AlertPresentation {
    let severity = (alert.severity != "Unknown").then(|| alert.severity.clone());
    let urgency = (alert.urgency != "Unknown").then(|| alert.urgency.clone());
    let areas: Vec<&String> = alert.areas.iter().take(3).collect();
    let regional = is_pirate_weather_alert(alert);

    let expires = alert.expires.map(|e| {
        format_display_datetime(
            &PyDateTime::aware(e, location_zone),
            &settings.time_display_mode,
            settings.time_format_12hour,
            settings.show_timezone_suffix,
            "%m/%d",
        )
    });
    let description = Some(alert.description.as_str())
        .filter(|d| !d.is_empty())
        .map(|d| truncate(d, 200));
    let instruction = alert
        .instruction
        .as_deref()
        .filter(|i| !i.is_empty())
        .map(|i| truncate(i, 150));

    let mut header = if alert.title.is_empty() {
        format!("Alert {index}")
    } else {
        format!("Alert {index}: {}", alert.title)
    };
    if let Some(label) = lifecycle_label.filter(|l| !l.is_empty()) {
        header = format!("{header} ({label})");
    }
    let mut parts = vec![header];
    let nonempty = |s: &Option<String>| s.clone().filter(|s| !s.is_empty());
    let mut bits = Vec::new();
    if let Some(s) = nonempty(&severity) {
        bits.push(format!("Severity: {s}"));
    }
    if let Some(u) = nonempty(&urgency) {
        bits.push(format!("Urgency: {u}"));
    }
    if !bits.is_empty() {
        parts.push(format!("  {}", bits.join(", ")));
    }
    let event = alert.event.as_deref().filter(|e| !e.is_empty());
    if let Some(e) = event {
        parts.push(format!("  Event: {e}"));
    }
    if !areas.is_empty() {
        let remaining = alert.areas.len() - areas.len();
        let mut text = areas
            .iter()
            .map(|a| a.as_str())
            .collect::<Vec<_>>()
            .join(", ");
        if remaining > 0 {
            text.push_str(&format!(" and {remaining} more"));
        }
        let label = if regional { "Regions" } else { "Areas" };
        parts.push(format!("  {label}: {text}"));
    }
    if regional {
        parts.push(
            "  Coverage: Regional alert from Pirate Weather/WMO; may not match your exact county or zone."
                .into(),
        );
    }
    if let Some(e) = &expires {
        parts.push(format!("  Expires: {e}"));
    }
    if let Some(d) = &description {
        parts.push(format!("  Description: {}", wrap_text(d, 80)));
    }
    if let Some(i) = &instruction {
        parts.push(format!("  Instructions: {}", wrap_text(i, 80)));
    }

    let title = if !alert.title.is_empty() {
        alert.title.clone()
    } else {
        event
            .map(str::to_string)
            .unwrap_or_else(|| format!("Alert {index}"))
    };
    AlertPresentation {
        title,
        severity,
        urgency,
        event: alert.event.clone(),
        areas: alert.areas.clone(),
        expires,
        description,
        instructions: instruction,
        fallback_text: parts.join("\n"),
    }
}
